### ISSUE TITLE
Hydra support for purpose on lights

### ISSUE THREAD

**Issue by rasmusbonnedal** [PixarAnimationStudios/USD#3059]
### Description
We have a `DomeLight` with a humongous environment texture like this:

```
def DomeLight "dome_environment" (
	prepend apiSchemas = ["ShadowAPI"]
)
{
	asset inputs:texture:file = @ljus1_ikea tom studio utan ljus_mono.hdr@
}
```

which causes the NVidia driver to throw up when loading the scene in usdview / Storm.
![nvidia_opengl_error](https://github.com/PixarAnimationStudios/OpenUSD/assets/4532327/6cc8d7ef-4d9a-4075-a45b-542fbd4dabe1)

We would like to have a preview version of the texture and @spiffmon told me the purpose token should work but is not supported by Hydra yet. It would be ideal for us to be able to author purpose on the lights like this and have it work in usdview:

```
def DomeLight "dome_environment" (
	prepend apiSchemas = ["ShadowAPI"]
)
{
	asset inputs:texture:file = @ljus1_ikea tom studio utan ljus_mono.hdr@
	uniform token purpose = "render"
}
```

### System Information (OS, Hardware)
Windows 10

### Package Versions
OpenUSD 23.11

**Comment by meshula** [comment]
@rasmusbonnedal When Hio opens images it's possible to resize to maximum possible. Apparently it's not getting limited upon load. How big is your HDR image? Could you share it with us for testing? It could be that we're trying to create a too-big image, it's also possible that the HDR image itself is being misparsed. And it's also possible that we don't attempt to limit a properly parsed incredibly large image to a plausible dimension.

**Comment by jesschimein** [comment]
Filed as internal issue #USD-9594

**Comment by rasmusbonnedal** [comment]
@meshula
```
>iinfo "ljus1_ikea tom studio utan ljus_mono.hdr"
ljus1_ikea tom studio utan ljus_mono.hdr : 5400 x 2700, 3 channel, float hdr
```

I generated a new image with the same specs and the problem is the same:
```
>oiiotool --pattern checker 5400x2700 3 -d float -o checker.hdr
```

I have attached a repro scene and my system information.

[NVIDIA System Information 04-29-2024 08-44-11.txt](https://github.com/PixarAnimationStudios/OpenUSD/files/15145873/NVIDIA.System.Information.04-29-2024.08-44-11.txt)

[repro_opengl_tdr_large_texture.zip](https://github.com/PixarAnimationStudios/OpenUSD/files/15145874/repro_opengl_tdr_large_texture.zip)

**Comment by fsuarezg** [comment]
Hello,

By any chance, is there an update on this ticket? We are encountering the same problem I believe with 2 dome lights, one with a large tx file and another with a smaller jpeg. Unfortunately, both get rendered and shown in the viewport (usdview) regardless of the purpose.
We are using OpenUSD 22.11 though so even older than the original poster (due to mayausd 0.30.0 in maya 2024).

Is there a workaround that we could do to make it work? I also tried placing the domelights under a transform with render purpose set, but unfortunately no luck.

Kind regards,
Fernando

**Comment by spiffmon** [comment]
Hi all,
   Even though the Lux lights _are_ Imageable, and thus endowed with `purpose`, internal discussion has concluded that we're not ready to make the assertion that purpose should be meaningful for lights.

   @rasmusbonnedal , would it be OK if we used this ticket and the repro you provided to investigate our texture handling and ensure that it do something reasonable?  Sounds like that may not be sufficient for @fsuarezg 's concerns; however we do have precedent in (I believe) our glsl texture reader to provide instruction on size/memory constraints to govern handling of large textures.  That does mean more up-front processing, however a mip-bearing texture format like OpenEXR (which has builtin support in Hio/OpenUSD) can mitigate that.

**Comment by fsuarezg** [comment]
Hi spiffmon, 

Thank you for your reply. I am interested in your comment:

> we're not ready to make the assertion that purpose should be meaningful for lights.

I can understand that in most cases lights would not really need a `purpose`, and in my specific case it is actually not really the light itself that I want to control the `purpose` of, but the image texture it uses. However doesn't light not having `purpose` working sort of defy the logic of it being an imageable object? or what are the consequences that you are worried about with lights having a purpose? 
I'm not sure if I have the wrong idea here, but I could imagine lights having some sort of viewport proxy representation like a cone for a spotlight.

Anyway, I would also be very happy with some sort of texture control on the attribute.

Maybe also interesting for anyone else that comes across this ticket: currently we have a workaround with a variantset called 'texture' where we have two options: 'jpeg'/'tx'. By default the 'jpeg' variant is always loaded, but when we render we have a pre-render script that forces the variant to 'tx'. It's not ideal, as you can still accidentally render the jpeg file, but it at least forces the right tx file on our farm.

**Comment by rasmusbonnedal** [comment]
@spiffmon Sure!

tl;dr
The issue is still present in 24.08 but we are not affected anymore.

We now route our envlight texture through a vray texture filter so Storm ignores the texture, so we haven't seen this issue in a while. But I tried to repro the issue this morning and I couldn't. The following variables have changed:

1. USD 24.08
2. New laptop with RTX 2000 / 8GB dedicated memory
3. NVidia driver 553.05, (still Windows 10)

So of course I tried to break it and upped the resolution to 12k x 6k and now the issue is there again. Same result with hdr, non-tiled exr and tx.

Command line to create texture:

```
> oiiotool --pattern checker 12000x6000 3 -d float -o checker.exr
> maketx checker.exr
```

The stage contains only this:
```
def DomeLight "dome_environment" (
	prepend apiSchemas = ["ShadowAPI"]
)
{
	asset inputs:texture:file = @checker.tx@
}
```

**Comment by spiffmon** [comment]
@fsuarezg ,
> However doesn't light not having `purpose` working sort of defy the logic of it being an imageable object? or what are the consequences that you are worried about with lights having a purpose?

Took some time, but we've cleared up some of the concerns, here.  So, you are correct, of course, and we agree lights should observe `purpose` like any other Imageable schema.  However, we are unsure whether our artists would want to use the feature (e.g. there _has_ been requests for official support for analytic proxy shapes for MeshLights without requiring a second prim to represent the proxy, but no requests for `proxy` purpose lights).  Therefore scheduling the project is tricky.  But if anyone in the community is interested in implementing the feature in Hydra 2, we'd be happy to consider a PR!

Question for you and @rasmusbonnedal ... if there were official support on textured lights for limiting the amount of memory consumed by the in-process texture (which would result in mip selection or resampling), would that be an acceptable solution, or is specifying an alternate texture (for specific renderers?  for proxy vs full purpose?) a superior, desired solution.  (And actually, even in the memory-limiting case, how would you want to specify the contexts in which you want to limit memory consumption, or is it OK to be unequivocal, or somehow associated with a RenderSettings prim?

**Comment by fsuarezg** [comment]
@spiffmon,

For me personally, I don't necessarily need to have some kind of automatic resampling happening to reduce memory usage. In my opinion - since this is dependent on the file type used within a textured light (jpeg, tx, hdr, png, ...) - it could lead to strange/unexpected behaviour and is too low level. 
Therefore, I prefer specifying an alternate file to use through the purpose. I think this is in line with the USD purpose philosophy. 

> But if anyone in the community is interested in implementing the feature in Hydra 2, we'd be happy to consider a PR!

I could have a look in my free time :)
